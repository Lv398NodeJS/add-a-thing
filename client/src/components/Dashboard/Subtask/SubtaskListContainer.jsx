import React, { Component } from 'react';
import { Row, Col, Container } from 'react-bootstrap';
import { connect } from 'react-redux';
import SubtaskProgressBar from './SubtaskProgressBar/SubtaskProgressBar';
import SubtaskList from './SubtaskList/SubtaskList';
import SubtaskFilterContainer from './SubtaskFilter/SubtaskFilterContainer';
import SubtaskAdd from './SubtaskAdd/SubtaskAdd';

export class SubtaskListContainer extends Component {
  render() {
    const {
      taskId,
      taskStatus,
      subtaskList = {},
      currentFilter,
    } = this.props;
    return (
      <Container fluid className="my-0 mx-0">
        <Row>
          <Col className="col-4">
            <h5>Subtasks:</h5>
          </Col>
          <Col className="col-8 text-right">
            <SubtaskFilterContainer />
          </Col>
        </Row>
        <SubtaskProgressBar subtaskList={subtaskList} />
        <SubtaskList
          taskId={taskId}
          taskStatus={taskStatus}
          subtaskList={subtaskList}
          currentFilter={currentFilter}
        />
        <SubtaskAdd taskId={taskId} taskStatus={taskStatus} />
      </Container>
    );
  }
}

const mapStateToProps = (
  {
    subtaskReducer: { subtaskList, currentFilter },
    taskDetailsReducer: { taskDetails: { _id: taskId, status: taskStatus } },
  },
) => ({
  subtaskList, currentFilter, taskId, taskStatus,
});

export { SubtaskListContainer as SubtaskListContainerComponent };
export default connect(
  mapStateToProps,
  null,
)(SubtaskListContainer);
