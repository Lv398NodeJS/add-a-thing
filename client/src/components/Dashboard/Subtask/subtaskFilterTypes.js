const SHOW_ALL = 'SHOW_ALL';
const SHOW_ACTIVE = 'SHOW_ACTIVE';
const SHOW_COMPLETED = 'SHOW_COMPLETED';

export {
  SHOW_ALL,
  SHOW_ACTIVE,
  SHOW_COMPLETED,
};
