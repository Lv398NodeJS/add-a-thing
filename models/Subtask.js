const mongoose = require('mongoose');

const { Schema } = mongoose;
const SubtaskSchema = new Schema({
  name: {
    type: String,
  },
  completed: Boolean,
  taskId: {
    type: Schema.Types.ObjectId,
    ref: 'Task',
    required: true,
  },
  date: {
    type: Date,
    default: Date.now,
  },
});

module.exports = Subtask = mongoose.model('Subtask', SubtaskSchema);
