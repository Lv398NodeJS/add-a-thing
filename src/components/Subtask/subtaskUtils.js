import { SHOW_ALL, SHOW_ACTIVE, SHOW_COMPLETED } from './subtaskFilterTypes';

export const getSubtaskListAsArray = (snapValue = {}) => {
  if (typeof snapValue !== 'object' && snapValue.constructor !== Object) return [];
  const subtaskList = [];
  Object.keys(snapValue).forEach((subtask) => {
    const { text, completed } = snapValue[subtask];
    subtaskList.push({
      text,
      completed,
      id: subtask,
    });
  });
  return subtaskList;
};

export const getVisibleSubtasks = (subtaskList, currentFilter) => {
  switch (currentFilter) {
    case SHOW_ALL:
      return subtaskList;
    case SHOW_COMPLETED:
      return subtaskList.filter(subtask => subtask.completed);
    case SHOW_ACTIVE:
      return subtaskList.filter(subtask => !subtask.completed);
    default:
      throw new Error(`Unknown filter: ${currentFilter}`);
  }
}