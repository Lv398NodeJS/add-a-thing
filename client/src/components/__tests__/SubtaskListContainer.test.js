import React from 'react';
import { shallow } from 'enzyme';
import { SubtaskListContainerComponent } from '../Dashboard/Subtask/SubtaskListContainer';

describe('SubtaskListContainer component', () => {
  const subtaskListContainerComponent = shallow(
    <SubtaskListContainerComponent />,
  );

  it('should render correctly and match the snapshot', () => {
    expect(subtaskListContainerComponent).toMatchSnapshot();
  });

  it('should render Bootstrap 4 Container component', () => {
    expect(subtaskListContainerComponent.find('Container').length).toBe(1);
  });

  it('should render Bootstrap 4 Row component', () => {
    expect(subtaskListContainerComponent.find('Bootstrap(Row)').length).toBe(1);
  });

  it('should render two Bootstrap 4 Col components', () => {
    expect(subtaskListContainerComponent.find('Col').length).toBe(2);
  });

  it('should render SubtaskProgressBar child component', () => {
    expect(subtaskListContainerComponent.find('SubtaskProgressBar').length).toBe(1);
  });

  it('should render SubtaskFilterContainer child component', () => {
    expect(subtaskListContainerComponent.find('SubtaskFilterContainer').length).toBe(1);
  });
});
