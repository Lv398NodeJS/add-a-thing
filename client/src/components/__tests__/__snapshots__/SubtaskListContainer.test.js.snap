// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubtaskListContainer component should render correctly and match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SubtaskListContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "as": "div",
      "children": Array [
        <ForwardRef(render)>
          <ForwardRef
            as="div"
            className="col-4"
          >
            <h5>
              Subtasks:
            </h5>
          </ForwardRef>
          <ForwardRef
            as="div"
            className="col-8 text-right"
          >
            <SubtaskFilterContainer />
          </ForwardRef>
        </ForwardRef(render)>,
        <SubtaskProgressBar
          subtaskList={Object {}}
        />,
        <UNDEFINED />,
        <UNDEFINED />,
      ],
      "className": "my-0 mx-0",
      "fluid": true,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <ForwardRef
              as="div"
              className="col-4"
            >
              <h5>
                Subtasks:
              </h5>
            </ForwardRef>,
            <ForwardRef
              as="div"
              className="col-8 text-right"
            >
              <SubtaskFilterContainer />
            </ForwardRef>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "as": "div",
              "children": <h5>
                Subtasks:
              </h5>,
              "className": "col-4",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Subtasks:",
              },
              "ref": null,
              "rendered": "Subtasks:",
              "type": "h5",
            },
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "as": "div",
              },
              "displayName": "Col",
              "render": [Function],
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "as": "div",
              "children": <SubtaskFilterContainer />,
              "className": "col-8 text-right",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "as": "div",
              },
              "displayName": "Col",
              "render": [Function],
            },
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": undefined,
          "displayName": "Bootstrap(Row)",
          "propTypes": undefined,
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "subtaskList": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "taskId": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(SubtaskList)",
          "type": [Function],
          Symbol(Symbol.hasInstance): [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "taskStatus": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(SubtaskAdd)",
          "type": [Function],
          Symbol(Symbol.hasInstance): [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "as": "div",
        "fluid": false,
      },
      "displayName": "Container",
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "as": "div",
        "children": Array [
          <ForwardRef(render)>
            <ForwardRef
              as="div"
              className="col-4"
            >
              <h5>
                Subtasks:
              </h5>
            </ForwardRef>
            <ForwardRef
              as="div"
              className="col-8 text-right"
            >
              <SubtaskFilterContainer />
            </ForwardRef>
          </ForwardRef(render)>,
          <SubtaskProgressBar
            subtaskList={Object {}}
          />,
          <UNDEFINED />,
          <UNDEFINED />,
        ],
        "className": "my-0 mx-0",
        "fluid": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <ForwardRef
                as="div"
                className="col-4"
              >
                <h5>
                  Subtasks:
                </h5>
              </ForwardRef>,
              <ForwardRef
                as="div"
                className="col-8 text-right"
              >
                <SubtaskFilterContainer />
              </ForwardRef>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "as": "div",
                "children": <h5>
                  Subtasks:
                </h5>,
                "className": "col-4",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Subtasks:",
                },
                "ref": null,
                "rendered": "Subtasks:",
                "type": "h5",
              },
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "as": "div",
                },
                "displayName": "Col",
                "render": [Function],
              },
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "as": "div",
                "children": <SubtaskFilterContainer />,
                "className": "col-8 text-right",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "as": "div",
                },
                "displayName": "Col",
                "render": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "displayName": "Bootstrap(Row)",
            "propTypes": undefined,
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "subtaskList": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "taskId": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.memo),
            "WrappedComponent": [Function],
            "compare": null,
            "displayName": "Connect(SubtaskList)",
            "type": [Function],
            Symbol(Symbol.hasInstance): [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "taskStatus": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.memo),
            "WrappedComponent": [Function],
            "compare": null,
            "displayName": "Connect(SubtaskAdd)",
            "type": [Function],
            Symbol(Symbol.hasInstance): [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "as": "div",
          "fluid": false,
        },
        "displayName": "Container",
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
